lib/resourcemerge/appt_test.go

```diff
@@ -0,0 +1,113 @@
+	daemonset_update_strategy_tests := []struct {
+		existing appsv1.DaemonSet
+		input    appsv1.DaemonSet
+
+		expectedModified bool
+		expected         appsv1.DaemonSet
+	}{
+		{
+			existing: appsv1.DaemonSet{
+				Spec: appsv1.DaemonSetSpec{
+					UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+						Type: "OldThing",
+						RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+							MaxUnavailable: &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+							MaxSurge:       &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+						},
+					},
+				},
+			},
+			input: appsv1.DaemonSet{
+				Spec: appsv1.DaemonSetSpec{
+					UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+						Type: "OldThing",
+						RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+							MaxUnavailable: &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+							MaxSurge:       &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+						},
+					},
+				},
+			},
+			//this "true" is expected true because of the nil pointers elsewhere in the DaemonSet struct
+			//it always takes new struct and sets modified to "true" in event of nil
+			expectedModified: true,
+			expected: appsv1.DaemonSet{
+				Spec: appsv1.DaemonSetSpec{
+					UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+						Type: "OldThing",
+						RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+							MaxUnavailable: &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+							MaxSurge:       &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+						},
+					},
+				},
+			},
+		}, {
+			existing: appsv1.DaemonSet{
+				Spec: appsv1.DaemonSetSpec{
+					UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+						Type: "OldThing",
+						RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+							MaxUnavailable: &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+							MaxSurge:       &intstr.IntOrString{Type: 0, IntVal: 1, StrVal: ""},
+						},
+					},
+				},
+			},
+			input: appsv1.DaemonSet{
+				Spec: appsv1.DaemonSetSpec{
+					UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+						Type: "NewThing",
+						RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+							MaxUnavailable: &intstr.IntOrString{Type: 0, IntVal: 2, StrVal: ""},
+							MaxSurge:       &intstr.IntOrString{Type: 0, IntVal: 3, StrVal: ""},
+						},
+					},
+				},
+			},
+			expectedModified: true,
+			expected: appsv1.DaemonSet{
+				Spec: appsv1.DaemonSetSpec{
+					UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+						Type: "NewThing",
+						RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+							MaxUnavailable: &intstr.IntOrString{Type: 0, IntVal: 2, StrVal: ""},
+							MaxSurge:       &intstr.IntOrString{Type: 0, IntVal: 3, StrVal: ""},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for idx, test := range daemonset_update_strategy_tests {
+		t.Run(fmt.Sprintf("test#%d", idx), func(t *testing.T) {
+			modified := BoolPtr(false)
+			EnsureDaemonSet(modified, &test.existing, test.input)
+
+			if *modified != test.expectedModified {
```

Comment:
So because of the above comment, this will always be `true` right?

I think this is fine as is, but of curiosity how much of the daemonset object would we have to mock for this to work as expected? It looks like ObjectMeta needs labels and annotations, and the Selector needs to be populated. Would that be enough?

Reply:
Correct. 

ObjectMeta and Selector would be enough if it EnsureDaemonSet didn't also call ensurePodTemplateSpec(), but because it does we have to:

1.  populate ObjectMeta on the template (because it calls EnsureObjectMeta also)
2.  populate the container lists because ensurePodSpec() doesn't ever initialize the existingCurr pointer for either set of containers, so if Spec->Template->PodSpec->InitContainers or Spec->Template->PodSpec->Containers is an empty list, it ends up treating it just like a nil and always returning modified
3. Same as #2, but for Spec->Template->PodSpec->Volumes
4. populate other PodSpec fields because they are are also checked (NodeSelector, SecurityContext,Affinity,Priority, etc) 

I have the test code to mock all of it stashed, so if we really want to mock the whole thing I will, I just was worried that the next poor soul was going to have to hunt through all of that again like I did if the structure changed and they didn't know why the test started failing.